airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/implementor/FailSyncTask.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.task.implementor
+
+import io.airbyte.cdk.state.SyncManager
+import io.airbyte.cdk.task.DestinationTaskExceptionHandler
+import io.airbyte.cdk.task.ImplementorTask
+import io.airbyte.cdk.write.DestinationWriter
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+import java.util.concurrent.atomic.AtomicBoolean
+
+interface FailSyncTask : ImplementorTask
+
+/**
+ * FailSyncTask is a task that is executed when a sync fails. It is responsible for cleaning up
+ * resources and reporting the failure.
+ */
+class DefaultFailSyncTask(
+    private val exceptionHandler: DestinationTaskExceptionHandler<*>,
+    private val destinationWriter: DestinationWriter,
+    private val exception: Exception,
+    private val syncManager: SyncManager,
+) : FailSyncTask {
+    private val log = KotlinLogging.logger {}
+    private val teardownRan = AtomicBoolean(false)
+
+    override suspend fun execute() {
+        if (teardownRan.compareAndSet(false, true)) {
```

Comment:
nit: I feel like we do this enough to warrant an extension function?
```kotlin
fun AtomicBoolean.oneshot(f: () -> Unit) {
  if (cas(false, true)) f()
}
```

or maybe just an actual Oneshot class
```kotlin
class Oneshot {
  private val flag = AtomicBoolean(false)
  fun execute(val f: () -> Unit) {
    if (flag.cas(false, true)) f()
  }
}
```

Reply:
Yeah, Iike that. Maybe `AtomicBoolean.setOnce...`?

Reply:
works for me :shrug:

Reply:
Done.